### PR TITLE
Emit TS declarations from JSDoc

### LINKED ISSUE #43: typescript definitions
hey @flekschas!
Great library! I was wondering if you already have some typescript definitions lying around? Otherwise I will try and write some
Thank you and have a great week

### PR DESCRIPTION
**Pull request by manzt**
Fixes #43 

As of TS 3.7 you can emit declarations `.d.ts` from JavaScript using JSDoc syntax. This PR adds JSDoc annotations using TS syntax to `src/index.js`. I did my best to capture what is described in the table in the project `README.md`, but some of the types might need refinement.

The TypeScript language server automatically parses the JSDoc annotations when working in `src/index.js`, offering code-completion when working in the project `src`. We can generate declarations for end users by adding `typescript` as dependency + `tsconfig.json`, or simply running the CLI via `npx`:

```bash
npx typescript dist/regl-scatterplot.esm.js --declaration --allowJs --emitDeclarationOnly
```

This command generates `dist/regl-scatterplot.esm.d.ts`:

<details><summary>Expand</summary>

```typescript
export default createScatterplot;
export type Color = [number, number, number, number] | [number, number, number];
export type Category = 'category' | 'value1' | 'valueA' | 'valueZ' | 'z';
export type Value = 'value' | 'value2' | 'valueB' | 'valueW' | 'w';
export type DataEncoding = Category | Value;
export type PointDataEncoding = DataEncoding | 'inherit' | 'segment';
export type Key = 'alt' | 'cmd' | 'ctrl' | 'meta' | 'shift';
export type Action = 'lasso' | 'rotate' | 'merge' | 'panZoom';
export type LassoClearEvent = 'deselect' | 'lassoEnd';
export type Camera2D = Record<string, any>;
export type Settable = {
    backgroundColor: Color | string;
    backgroundImage: import('regl').Texture2D | string;
    colorBy: DataEncoding;
    deselectOnDblClick: boolean;
    deselectOnEscape: boolean;
    lassoColor: Color | string;
    lassoLineWidth: number;
    lassoMinDelay: number;
    lassoMinDist: number;
    lassoClearEvent: LassoClearEvent;
    lassoInitiator: boolean;
    keyMap: Record<Key, Omit<Action, 'panZoom'>>;
    mouseMode: Omit<Action, 'merge'>;
    showReticle: boolean;
    reticleColor: Color | string;
    pointColor: Color;
    pointColorActive: Color;
    pointColorHover: Color;
    showPointConnections: boolean;
    pointConnectionColor: Color;
    pointConnectionColorActive: Color;
    pointConnectionColorHover: Color;
    pointConnectionColorBy: PointDataEncoding;
    pointConnectionOpacity: number;
    pointConnectionOpacityBy: PointDataEncoding;
    pointConnectionOpacityActive: number;
    pointConnectionSize: number;
    pointConnectionSizeActive: number;
    pointConnectionSizeBy: PointDataEncoding;
    pointConnectionMaxIntPointsPerSegment: number;
    pointConnectionTolerance: number;
    pointSize: number;
    pointSizeSelected: number;
    pointSizeMouseDetection: 'auto';
    pointOutlineWidth: number;
    opacity: number;
    opacityBy: DataEncoding;
    opacityByDensityFill: number;
    sizeBy: DataEncoding;
    height: 'auto' | number;
    width: 'auto' | number;
    gamma: number;
};
export type Properties = {
    canvas: HTMLElement;
    regl: import('regl').Regl;
    syncEvents: boolean;
    version: string;
    lassoInitiatorElement: HTMLElement;
    camera: Camera2D;
    performanceMode: boolean;
    opacityByDensityDebounceTime: number;
} & Settable;
/**
 * @typedef {[number, number, number, number] | [number, number, number]} Color
 *
 * @typedef {'category' | 'value1' | 'valueA' | 'valueZ' | 'z'} Category
 *
 * @typedef {'value' | 'value2' | 'valueB' | 'valueW' | 'w'} Value
 *
 * @typedef {Category | Value} DataEncoding
 *
 * @typedef {DataEncoding | 'inherit' | 'segment'} PointDataEncoding
 *
 * @typedef {'alt' | 'cmd' | 'ctrl' | 'meta' | 'shift'} Key
 *
 * @typedef {'lasso' | 'rotate' | 'merge' | 'panZoom'} Action
 *
 * @typedef {'deselect' | 'lassoEnd'} LassoClearEvent
 *
 * @typedef {Record<string, any>} Camera2D
 *
 * @typedef {{
 *   backgroundColor: Color | string;
 *   backgroundImage: import('regl').Texture2D | string;
 *   colorBy: DataEncoding;
 *   deselectOnDblClick: boolean;
 *   deselectOnEscape: boolean;
 *   lassoColor: Color | string;
 *   lassoLineWidth: number;
 *   lassoMinDelay: number;
 *   lassoMinDist: number;
 *   lassoClearEvent: LassoClearEvent;
 *   lassoInitiator: boolean;
 *   keyMap: Record<Key, Omit<Action, 'panZoom'>>;
 *   mouseMode: Omit<Action, 'merge'>;
 *   showReticle: boolean;
 *   reticleColor: Color | string;
 *   pointColor: Color;
 *   pointColorActive: Color;
 *   pointColorHover: Color;
 *   showPointConnections: boolean;
 *   pointConnectionColor: Color;
 *   pointConnectionColorActive: Color;
 *   pointConnectionColorHover: Color;
 *   pointConnectionColorBy: PointDataEncoding;
 *   pointConnectionOpacity: number;
 *   pointConnectionOpacityBy: PointDataEncoding;
 *   pointConnectionOpacityActive: number;
 *   pointConnectionSize: number;
 *   pointConnectionSizeActive: number;
 *   pointConnectionSizeBy: PointDataEncoding;
 *   pointConnectionMaxIntPointsPerSegment: number;
 *   pointConnectionTolerance: number;
 *   pointSize: number;
 *   pointSizeSelected: number;
 *   pointSizeMouseDetection: 'auto';
 *   pointOutlineWidth: number;
 *   opacity: number;
 *   opacityBy: DataEncoding;
 *   opacityByDensityFill: number;
 *   sizeBy: DataEncoding;
 *   height: 'auto' | number;
 *   width: 'auto' | number;
 *   gamma: number;
 * }} Settable
 *
 * @typedef {{
 *   canvas: HTMLElement;
 *   regl: import('regl').Regl;
 *   syncEvents: boolean;
 *   version: string;
 *   lassoInitiatorElement: HTMLElement;
 *   camera: Camera2D;
 *   performanceMode: boolean;
 *   opacityByDensityDebounceTime: number;
 * } & Settable} Properties
 *
 * @param {Partial<Properties>} initialProperties
 */
declare function createScatterplot(initialProperties?: Partial<Properties>): {
    clear: () => void;
    createTextureFromUrl: (url: any) => Promise<import("regl").Texture2D>;
    deselect: ({ preventEvent }?: {
        preventEvent?: boolean;
    }) => void;
    destroy: () => void;
    draw: (newPoints: number[][], options?: Partial<{
        transition: boolean;
        transitionDuration: number;
        transitionEasing: string;
    }>) => Promise<void>;
    get: <Key_1 extends "camera" | "canvas" | "regl" | "syncEvents" | "version" | "lassoInitiatorElement" | "performanceMode" | "opacityByDensityDebounceTime" | keyof Settable>(property: Key_1) => Properties[Key_1];
    hover: (point: number | number[], { showReticleOnce, preventEvent }?: {
        showReticleOnce?: boolean;
        preventEvent?: boolean;
    }) => void;
    refresh: () => void;
    reset: () => void;
    select: (pointIdxs: number | number[], { merge, preventEvent }?: {
        merge?: boolean;
        preventEvent?: boolean;
    }) => void;
    set: (properties: Partial<Settable>) => void;
    export: (options?: import('regl').ReadOptions<Uint8Array>) => {
        pixels: Uint8ClampedArray;
        width: number;
        height: number;
    };
    subscribe: <Event_1 extends keyof {
        init: () => void;
        destroy: () => void;
        backgroundImageReady: () => void;
        pointOver: (pointIndex: number) => void;
        pointOut: (pointIndex: number) => void;
        select: (payload: {
            points: number[][];
        }) => void;
        deselect: () => void;
        view: (payload: {
            camera: Camera2D;
            view: any;
            xScale: [number, number];
            yScale: [number, number];
        }) => void;
        draw: (payload: {
            camera: Camera2D;
            view: any;
            xScale: [number, number];
            yScale: [number, number];
        }) => void;
        lassoStart: () => void;
        lassoExtend: (payload: {
            coordinates: number[];
        }) => void;
        lassoEnd: (payload: {
            coordinates: number[];
        }) => void;
        transtionStart: () => void;
        transitionEnd: (payload: import('regl').Regl) => void;
        pointConnectionsDraw: () => void;
    }>(event: Event_1, handler: {
        init: () => void;
        destroy: () => void;
        backgroundImageReady: () => void;
        pointOver: (pointIndex: number) => void;
        pointOut: (pointIndex: number) => void;
        select: (payload: {
            points: number[][];
        }) => void;
        deselect: () => void;
        view: (payload: {
            camera: Camera2D;
            view: any;
            xScale: [number, number];
            yScale: [number, number];
        }) => void;
        draw: (payload: {
            camera: Camera2D;
            view: any;
            xScale: [number, number];
            yScale: [number, number];
        }) => void;
        lassoStart: () => void;
        lassoExtend: (payload: {
            coordinates: number[];
        }) => void;
        lassoEnd: (payload: {
            coordinates: number[];
        }) => void;
        transtionStart: () => void;
        transitionEnd: (payload: import('regl').Regl) => void;
        pointConnectionsDraw: () => void;
    }[Event_1], times: number) => void;
    unsubscribe: <Event_2 extends keyof {
        init: () => void;
        destroy: () => void;
        backgroundImageReady: () => void;
        pointOver: (pointIndex: number) => void;
        pointOut: (pointIndex: number) => void;
        select: (payload: {
            points: number[][];
        }) => void;
        deselect: () => void;
        view: (payload: {
            camera: Camera2D;
            view: any;
            xScale: [number, number];
            yScale: [number, number];
        }) => void;
        draw: (payload: {
            camera: Camera2D;
            view: any;
            xScale: [number, number];
            yScale: [number, number];
        }) => void;
        lassoStart: () => void;
        lassoExtend: (payload: {
            coordinates: number[];
        }) => void;
        lassoEnd: (payload: {
            coordinates: number[];
        }) => void;
        transtionStart: () => void;
        transitionEnd: (payload: import('regl').Regl) => void;
        pointConnectionsDraw: () => void;
    }>(event: Event_2) => void;
};
/**
 * Create a new Regl instance with `GL_EXTENSIONS` enables
 * @param   {HTMLCanvasElement}  canvas  Canvas element to be rendered on
 * @return  {import('regl').Regl}  New Regl instance
 */
export function createRegl(canvas: HTMLCanvasElement): import('regl').Regl;
/**
 * Create a Regl texture from an URL.
 * @param   {import('regl').Regl}  regl  Regl instance used for creating the texture.
 * @param   {string}  url  Source URL of the image.
 */
export function createTextureFromUrl(regl: import('regl').Regl, url: string): Promise<import("regl").Texture2D>;
declare var version: string;

```
</details>

I'll post some screenshots below of some nice in-editor intellisense.

_TODO_

- [x] : decide whether to add TS as a dependency or use `npx`

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -185,7 +185,115 @@ const getEncodingIdx = (type) => {
...
+ *   syncEvents: boolean;
+ *   version: string;
+ *   lassoInitiatorElement: HTMLElement;
+ *   camera: Camera2D;
+ *   performanceMode: boolean;
+ *   opacityByDensityDebounceTime: number;
+ * } & Settable} Properties
+ *
+ * @param {Partial<Properties>} initialProperties
+ */
```
Comment (manzt) [inline src/index.js:264]: Alternatively all types could be declared in `src/types.d.ts` and then imported in `src/index.js`:

```javascript
 *
 * @param {Partial<import('./types').Properties>} initialProperties
 */
```

File: src/index.js [hunk 2]
```diff
@@ -185,7 +185,115 @@ const getEncodingIdx = (type) => {
...
+   *   transtionStart: () => void;
+   *   transitionEnd: (payload: import('regl').Regl) => void;
+   *   pointConnectionsDraw: () => void;
+   * }} PubSubEvents
+   *
+   * @type {{
+   *  subscribe: <Event extends keyof PubSubEvents>(event: Event, handler: PubSubEvents[Event], times: number) => void;
+   *  unsubscribe: <Event extends keyof PubSubEvents>(event: Event) => void;
+   * }}
+   */
```
Comment (manzt) [inline src/index.js:296]: I'm rather happy with this bit. Based on the event name, TS will infer the signature of the callback:

image

File: src/index.js [hunk 3]
```diff
@@ -2228,6 +2351,7 @@ const createScatterplot = (initialProperties = {}) => {
     regl.poll();
   };
 
+  /** @type {<Key extends keyof Properties>(property: Key) => Properties[Key] } */
```
Comment (manzt) [inline src/index.js:2354]: Similar here, TS will infer the return type based on the property name:

- List all properties, can tab complete:

image


- Auto infer return type:

image

**Comment by manzt** [comment]
TS type inference is pretty strong, so JSDoc annotations can just be sprinkled in to make the output better. As is, the default export is essentially fully typed:

image

File: src/utils.js [hunk 4]
```diff
@@ -140,12 +141,12 @@ export const loadImage = (src, isCrossOrigin = false) =>
   });
 
 /**
+ * @deprecated Please use `scatterplot.createTextureFromUrl(url)`
+ *
```
Comment (manzt) [inline src/utils.js:145]: End users will get a deprecation notice in their text editor when importing this function.

Comment (manzt) [reply src/utils.js:145]: image

File: src/types.d.ts [hunk 5]
```diff
@@ -0,0 +1,148 @@
+type Color = [number, number, number, number] | [number, number, number];
```
Comment (manzt) [inline src/types.d.ts:1]: This is probably the type that needs the most help. Is string allowed for all color props as well, or just some?

File: src/types.d.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  reticleColor: Color;
+  opacity: number;
+  opacityByDensityFill: number;
+  height: 'auto' | number;
+  width: 'auto' | number;
+  gamma: number;
+} & PointOptions &
+  CameraOptions &
+  LassoOptions &
+  Nullifiable;
```
Comment (manzt) [inline src/types.d.ts:80]: Separates `Settable` properties from non-settable Properties.

When you call `plot.set()` only keys & values from this interface can be used.

File: src/types.d.ts [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+export type Properties = {
+  canvas: HTMLCanvasElement;
+  regl: import('regl').Regl;
+  syncEvents: boolean;
+  version: string;
+  lassoInitiatorElement: HTMLElement;
+  camera: Camera2D;
+  performanceMode: boolean;
+  opacityByDensityDebounceTime: number;
+} & Settable;
```
Comment (manzt) [inline src/types.d.ts:91]: When you call `plot.get` you can return any of the properties on this type, including Settable properties.

File: src/types.d.ts [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+  | 'lassoStart'
+  | 'transitionStart'
+  | 'pointConnectionsDraw',
+  undefined
+> &
+  PubSubEvent<'view' | 'draw', DrawPayload> &
+  PubSubEvent<'lassoEnd' | 'lassoExtend', { coordinates: number[] }> &
+  PubSubEvent<'pointOver' | 'pointOut', number> &
+  PubSubEvent<'points', { points: number[][] }> &
+  PubSubEvent<'transitionEnd', import('regl').Regl>;
```
Comment (manzt) [inline src/types.d.ts:135]: Maybe a little confusing, but I think the easiest way to define or extend the behavior of the custom pubsub regl-scatter wraps. Let me know if you have any questions on extending this.

**Comment by manzt** [comment]
@flekschas should be good for review. My suggestion would be to pull down this branch and try out the following:

```bash
$ npm run build-library
$ npm pack

$ mkdir test-regl && cd test-regl
$ npm init --y
$ npm install ../regl-scatterplot-0.19.2.tgz
$ touch index.js # try importing 'regl-scatterplot'
```

File: src/utils.js (hunk 4 above)
Comment (flekschas) [reply src/utils.js:145]: 😍

**Review by flekschas — APPROVED** [review]
Looks fantastic! Thanks a lot for creating the PR! If you have the chance, can you test the modified color definition?

**Comment by manzt** [comment]
Tried it out! Thanks for updating.

**Pull request merged**